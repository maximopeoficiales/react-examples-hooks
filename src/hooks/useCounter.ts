import { useState } from "react"

export const useCounter = (numberInitial: number = 200) => {
    const [counter, setCounter] = useState(numberInitial);

    const increment = (factor: number = 1) => {
        if (counter <= 0) {
            setCounter(numberInitial)
        } else {
            setCounter(counter + factor)
        }
    }
    const decrement = (factor: number = 1) => {
        setCounter(counter - factor)
    }

    const reset = () => {
        setCounter(numberInitial);
    }

    return {
        counter,
        increment, decrement, reset
    }
}
