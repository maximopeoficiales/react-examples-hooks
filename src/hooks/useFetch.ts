import { useEffect, useRef, useState } from "react";

interface UseFecthData {
    data: any,
    loading: boolean,
    error: any
}
export const useFetch = (url: string) => {
    // uso de useRef
    const isMounted = useRef(true);
    const [state, setState] = useState(
        {
            data: null, loading: true, error: null
        } as UseFecthData
    );

    useEffect(() => {

        // cuando se desmonte el componente la referencia estara en false
        return () => {

            // isMounted.current = false
        }
    }, []);

    useEffect(() => {
        setState({ loading: true, data: null, error: null });

        (async () => {
            // if (isMounted.current) {
            let data = await (await fetch(url)).json();
            setState({ loading: false, data, error: null });
            // }else{
            // console.log("El useState no se esta ejecutando ");

            // }
        })()
    }, [url])
    return { state };
}
