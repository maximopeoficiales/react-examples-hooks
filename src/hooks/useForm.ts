import { ChangeEvent, useState } from "react";

export function useForm<Type>(values: Type) {
    const [formValues, setFormValues] = useState(values);

    const handlerChange = ({ target }: ChangeEvent<HTMLInputElement>) => {
        setFormValues({
            ...formValues,
            [target.name]: target.value,
        });
    };

    return { formValues, handlerChange };


}
