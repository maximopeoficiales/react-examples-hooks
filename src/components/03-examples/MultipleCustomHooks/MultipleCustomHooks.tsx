import { useCounter } from "../../../hooks/useCounter";
import { useFetch } from "../../../hooks/useFetch";

interface MyProps {}
const defaultProps: MyProps = {};
const MultipleCustomHooks = (props: MyProps) => {
  props = { ...defaultProps, ...props };

  const { counter, increment, decrement } = useCounter(1);
  const { state } = useFetch(
    `https://www.breakingbadapi.com/api/characters/${counter}`
  );
  // console.log(state);
  const { loading, data } = state;
  // cuando sea diferente de null obten este dato
  const { name, birthday, img, status } = !!data && data[0];

  return (
    <div
      data-testid="MultipleCustomHooks"
      className="row justify-content-center"
    >
      <div className="col-md-4">
        <h1>Breaking Bad </h1>
        {loading ? (
          <h5>Loading.....</h5>
        ) : (
          <>
            <div className="card">
              <img className="card-img-top " src={img} alt="" height="400px" />
              <div className="card-body">
                <h4 className="card-title">{name}</h4>
                <p className="card-text">{birthday}</p>
                <p className="card-text">{status}</p>
              </div>
            </div>
            <div className="d-flex justify-content-center my-2">
              <button
                className="btn btn-lg btn-primary mx-2"
                onClick={() => {
                  decrement();
                }}
              >
                Previous
              </button>

              <button
                className="btn btn-lg btn-primary mx-2"
                onClick={() => {
                  increment();
                }}
              >
                Next
              </button>
            </div>
          </>
        )}
      </div>
    </div>
  );
};

export default MultipleCustomHooks;
