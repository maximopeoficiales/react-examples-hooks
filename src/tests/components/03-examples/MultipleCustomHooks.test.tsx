import { shallow } from "enzyme";
import React from "react";
import MultipleCustomHooks from "../../../components/03-examples/MultipleCustomHooks/MultipleCustomHooks";
import { useCounter } from "../../../hooks/useCounter";
import { UseFecthData, useFetch } from "../../../hooks/useFetch";
// simula comportamientos de los resultados de los hooks
jest.mock("../../../hooks/useCounter");
jest.mock("../../../hooks/useFetch");

describe("Pruebas en MultipleCustomHooks", () => {
  const stateCounter = {
    counter: 1,
    increment: () => {},
    decrement: () => {},
    reset: () => {},
  };

  const state: UseFecthData = {
    data: null,
    loading: true,
    error: null,
  };

  test("debe de mostrarse correctamente", () => {
    (useCounter as jest.Mock).mockReturnValue({ stateCounter });

    (useFetch as jest.Mock).mockReturnValue({ state });
    const wrapper = shallow(<MultipleCustomHooks />);

    expect(wrapper).toMatchSnapshot();
  });

  test("should debe de mostrar la info", () => {
    // simulo que ya cargo los datos
    (useCounter as jest.Mock).mockReturnValue({ stateCounter });

    (useFetch as jest.Mock).mockReturnValue({
      data: [
        { name: "maximo", birthday: "laksjdlk", img: "asd", status: "safd" },
      ],
      error: null,
      loading: false,
    });
    const wrapper = shallow(<MultipleCustomHooks />);
    expect(wrapper.find(".card-title").text().trim()).toBe("maximo");
    expect(wrapper.find("h5").exists()).toBe(false);

    // console.log(wrapper.html());

    // expect(wrapper).toMatchSnapshot();
  });
});
