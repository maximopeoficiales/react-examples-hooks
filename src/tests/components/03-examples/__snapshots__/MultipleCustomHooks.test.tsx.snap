// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en MultipleCustomHooks debe de mostrarse correctamente 1`] = `
<div
  className="row justify-content-center"
  data-testid="MultipleCustomHooks"
>
  <div
    className="col-md-4"
  >
    <h1>
      Breaking Bad 
    </h1>
    <div
      className="card"
    >
      <img
        alt=""
        className="card-img-top "
        height="400px"
      />
      <div
        className="card-body"
      >
        <h4
          className="card-title"
        />
        <p
          className="card-text"
        />
        <p
          className="card-text"
        />
      </div>
    </div>
    <div
      className="d-flex justify-content-center my-2"
    >
      <button
        className="btn btn-lg btn-primary mx-2"
        onClick={[Function]}
      >
        Previous
      </button>
      <button
        className="btn btn-lg btn-primary mx-2"
        onClick={[Function]}
      >
        Next
      </button>
    </div>
  </div>
</div>
`;
